File: RayTracer/Ray.fs
module RayTracer.Ray

open System.Numerics
open Color

type Ray =
    { origin: Vector3
      direction: Vector3
      color: SpectralColor
      reflectionRemains : uint }

module Ray =
    let at (ray: Ray) (distance: float32): Vector3 = ray.direction * distance + ray.origin

    let reflect (ray: Ray) (normal: Vector3) (point: Vector3) (colorFilter: SpectralColor): Ray =
        { origin = point
          direction = Vector3.Reflect(ray.direction, normal)
          color = Array.map2 (*) (ray.color.values) (colorFilter.values) |> SpectralColor.initFromArray
          reflectionRemains = ray.reflectionRemains - 1u}

File: RayTracerCs/Ray.cs
using System.Numerics;

namespace RayTracerCs
{
    public struct Ray
    {
        public Ray(Vector3 origin, Vector3 direction, SpectralColor color)
        {
            Origin = origin;
            Direction = direction;
            ColorFilter = color;
        }

        public Vector3 Origin { get; }
        public Vector3 Direction { get; }
        public SpectralColor ColorFilter { get; }
        public Vector3 At(float distance) => Origin + Direction * distance;

        public Ray Reflect(Vector3 normal, Vector3 point, SpectralColor colorFilter)
        {
            return new Ray(
                point,
                Vector3.Reflect(Direction, normal),
                ColorFilter.ApplyFilter(colorFilter)
            );
        }
    }
}